db/migrate/20200813175255_set_team_fieldsets.rb

```diff
@@ -0,0 +1,53 @@
+class SetTeamFieldsets < ActiveRecord::Migration
+  def change
+    started = Time.now.to_i
+    RequestStore.store[:skip_notifications] = true
+    fieldsets = [
+      {
+        identifier: 'tasks',
+        singular: 'task',
+        plural: 'tasks'
+      }.with_indifferent_access,
+      {
+        identifier: 'metadata',
+        singular: 'metadata',
+        plural: 'metadata'
+      }.with_indifferent_access
+    ]
+    # Set fieldsets setting to be the default for all teams
+    Team.find_each do |t|
+      settings = t.settings || {}
+      settings[:fieldsets] = fieldsets
+      t.update_column(:settings, settings)
+    end
+    # Set fieldset = "tasks" for all TeamTasks
+    TeamTask.update_all(fieldset: 'tasks')
+    # Set fieldset: "tasks" for all Tasks
+    count = Task.where(annotation_type: 'task').count
+    if count > 0
+      # remove annotation indexes (to make migration faster)
+      remove_index :annotations, name: "index_annotations_on_annotation_type"
+      remove_index :annotations, name: "index_annotations_on_annotated_type_and_annotated_id"
+      remove_index :annotations, name: "index_annotation_type_order"
+      Task.where(annotation_type: 'task').find_in_batches(:batch_size => 10000) do |tasks|
+        new_tasks = []
+        tasks.each do |t|
+          print "."
+          t.data["fieldset"] = 'tasks'
+          new_tasks <<  t
+        end
+        # Delete existing task before import new tasks
+        Task.where(id: tasks.map(&:id)).delete_all
```

Comment:
This operation looks risky if something goes wrong between deleting and importing... did you try to force an exception between those lines (for example, add a `raise` to line 41) and see if the deleted tasks are reverted?

Reply:
Added a `raise` based on the return of `import` action

Reply:
Thanks! But the real question is: when `raise` runs and the whole migration is interrupted... the deleted tasks on line 44 are reverted? Or they are gone forever?

Reply:
They gone forever and print those tasks on console so we can take the action manually

Reply:
Hummm... still risky... let's discuss more on the call

Reply:
I can keep the old tasks and try the import with option `on_duplicate_key_update` but I think it'll take long time

Reply:
It's worth trying, at least to be sure of how long it takes.

Reply:
Yes I am trying now

Reply:
with `on_duplicate_key_update` options the migration took `62m` and I tested with/without indexed and the difference is `1m` to I ignore to delete/create indexes

Reply:
awesome Sawy, in that case, let's keep the `on_duplicate_key_update`

Reply:
OK